src/main/java/com/example/demo/services/BookingService.java

```diff
@@ -0,0 +1,272 @@
+
+        List<Long> reservedSeatIds = movie.getReservations().stream()
+                .map(ReservationEntity::getSeatId).collect(Collectors.toList());
+
+        List<SeatEntity> availableSeats = movie.getRoom().getSeats().stream()
+                .filter(item -> !reservedSeatIds.contains(item.getId()))
+                .collect(Collectors.toList());
+
+        List<Long> availableSeatIds = availableSeats.stream().map(SeatEntity::getId).collect(Collectors.toList());
+
+        // check if desired seats are empty
+        for (Pair<Long, String> seat : request.getSeats()) {
+            if (!availableSeatIds.contains(seat.getFirst())) {
+                throw new IllegalArgumentException("Seat is already taken");
+            }
+        }
+
+        List<Long> seatsFromReservation =
+                request.getSeats().stream()
+                        .map(Pair<Long, String>::getFirst)
+                        .collect(Collectors.toList());
+
+        // free after reservation
+        availableSeatIds.removeAll(seatsFromReservation);
+
+
+        //check if there exists a singular free seat between reserved seats
+
+        List<SeatEntity> sortedAllSeats = deepCopySeatList(movie.getRoom().getSeats());
+        Collections.sort(sortedAllSeats, new SortSeatEntity());
+
+        //all free seats have null id, taken have not null
+        sortedAllSeats.stream().map(item -> {
+            if (availableSeatIds.contains(item.getId())) item.setId(null);
+            return item;
+        }).collect(Collectors.toList());
+
+
+        for (int i = 1; i < sortedAllSeats.size() - 1; i++) {
+            SeatEntity left = sortedAllSeats.get(i - 1);
+            SeatEntity center = sortedAllSeats.get(i);
+            SeatEntity right = sortedAllSeats.get(i + 1);
+
+            //empty seat between reserved on sorted list, need to check rowNumber and positionInRow
+            if (center.getId() == null && right.getId() != null && left.getId() != null) {
+                if (Objects.equals(left.getRowNumber(), center.getRowNumber())
+                        && Objects.equals(center.getRowNumber(), right.getRowNumber())) {
+                    if (left.getPositionInRow() + 1 == center.getPositionInRow()
+                            && right.getPositionInRow() - 1 == center.getPositionInRow()) {
+                        throw new IllegalArgumentException(" single leftover seat");
+                    }
+                }
+            }
+        }
+
+        logger.debug("passed checks");
+        for (SeatEntity s : sortedAllSeats) {
+            logger.debug(s.toString());
+
+        }
+
+        //checks passed init entity and save
+        for (Long id : seatsFromReservation) {
+            ReservationEntity reservationEntity = reservationRepository.save(
+                    new ReservationEntity(request.getName(), request.getSurname(), id));
+            movie.getReservations().add(reservationEntity);
+        }
+        movieRepository.save(movie);
```

Comment:
Co się stanie jak dwóch użytkowników w tym samym momencie poprosi o to samo miejsce?

Reply:
z tego co wiem najpierw wykona się jedna transakcja a potem druga , natomiast wiem że temat jest głębszy (inne poziomy izolacji)  i powinienem się z nim lepiej zapoznać